FhirToDataLake/src/Microsoft.Health.Fhir.Synapse.Core/Tasks/TaskExecutor.cs

```diff
@@ -29,138 +36,475 @@ public sealed class TaskExecutor : ITaskExecutor
+
+                            var patientInfo = taskContext.Patients[patientIndex];
+
+                            // the patient resource isn't included in compartment search,
+                            // so we need additional request to get the patient resource
+
+                            // the patient resource is not retrieved yet,
+                            // for resumed task, the current patient may be retrieved and its version id exists.
+                            if (!cacheResult.SearchProgress.PatientVersionId.ContainsKey(patientInfo.PatientId))
+                            {
+                                var patientResource = await GetPatientResource(patientInfo, cancellationToken);
+
+                                // the patient does not exist
+                                if (patientResource == null)
+                                {
+                                    continue;
+                                }
+
+                                var versionId = FhirBundleParser.ExtractVersionId(patientResource);
+
+                                if (versionId == 0)
+                                {
+                                    _logger.LogError($"Failed to extract version id for patient {patientInfo.PatientId}.");
+                                    throw new FhirSearchException($"Failed to extract version id for patient {patientInfo.PatientId}.");
+                                }
+
+                                // New patient or the patient is updated.
+                                if (patientInfo.VersionId != versionId)
+                                {
+                                    // save the patient resource to cache if the patient resource type is required in the result
+                                    if (isPatientResourcesRequired)
+                                    {
+                                        AddFhirResourcesToCache(new List<JObject> { patientResource }, cacheResult);
+                                    }
+                                }
+
+                                // add this patient's version id in cacheResult,
+                                // the version id will be synced to taskContext when the cache result is committed, and be recorded in job/schedule metadata further
+                                cacheResult.SearchProgress.PatientVersionId[patientInfo.PatientId] = versionId;
+                                _logger.LogInformation($"Get patient resource {patientInfo.PatientId} successfully.");
+                            }
+
+                            // the version id is 0 for newly patient
+                            // for new patient, we will retrieve all its compartments resources from {since}
+                            // for processed patient, we will only retrieve the updated compartment resources from last scheduled time
+                            var startDateTime = patientInfo.VersionId == 0
```

Comment:
Maybe new patient can also have version id bigger than 0?

Reply:
The version id used here is the version id got from `schedulermetadata` instead of the version id extracted from fhir resource `patientResource`, it is set to 0 for a new patient.

I will refine the name for better distringuish.